packages/react/src/SelectPanel/SelectPanel.tsx

```diff
@@ -477,11 +482,34 @@ export function SelectPanel({
     }
   }
 
-  // We add a save and cancel button on narrow screens when SelectPanel is full-screen
-  const showCancelSaveButtons = variant === 'modal' || (isMultiSelectVariant(selected) && usingFullScreenOnNarrow)
   // because of instant selection, canceling on single select is the same as closing the panel, no onCancel needed
   const showXCloseIcon =
-    variant === 'modal' || ((onCancel || !isMultiSelectVariant(selected)) && usingFullScreenOnNarrow)
+    variant === 'modal' || ((onCancel !== undefined || !isMultiSelectVariant(selected)) && usingFullScreenOnNarrow)
+
+  // We add a save and cancel button on:
+  // - modals
+  // - anchored panels with multi select if there is onCancel
+  const showCancelSaveButtons =
+    variant === 'modal' || (isMultiSelectVariant(selected) && usingFullScreenOnNarrow && onCancel !== undefined)
+
+  // The responsive save button is only covering a very specific case:
+  // - anchored panel with multi select if there is no onCancel
+  const showResponsiveSaveButton =
+    variant !== 'modal' && usingFullScreenOnNarrow && isMultiSelectVariant(selected) && onCancel === undefined
+
+  // If there is any element in the footer, we render it.
+  const renderFooter = secondaryAction !== undefined || showCancelSaveButtons || showResponsiveSaveButton
+
+  // If there's any permanent elements in the footer, we show it always.
+  // The save and close button is only shown on small screens.
+  const displayFooter =
+    secondaryAction !== undefined || showCancelSaveButtons
+      ? 'always'
+      : showResponsiveSaveButton
+        ? 'only-small'
+        : undefined
+
+  const stretchSecondaryAction = showResponsiveSaveButton ? 'only-big' : showCancelSaveButtons ? 'never' : 'always'
```

Comment:
Rewrote all these conditions to make it easier to understand. It is still very complex and warrants a rewrite tbh, but this should be an improvement.